src/simplifiers/BoolRewriting.h

```diff
@@ -77,19 +77,34 @@ PTRef rewriteMaxArity(Logic & logic, const PTRef root, T doNotRewrite) {
             else if (logic.isAtom(t[i])) {
                 assert(!cache.has(t[i]));
                 cache.insert(t[i], t[i]);
+            } else {
+                assert(false);
             }
         }
         if (unprocessed_children)
             continue;
 
         unprocessed_ptrefs.pop();
-        PTRef result = PTRef_Undef;
-        assert(logic.isBooleanOperator(tr));
+        SymRef symRef = t.symb();
+        assert(logic.isBooleanOperator(symRef));
 
-        if (logic.isAnd(tr) || logic.isOr(tr)) {
+        PTRef result = PTRef_Undef;
+        if (logic.isAnd(symRef) or logic.isOr(symRef)) {
             result = ::mergeAndOrArgs(logic, tr, cache, doNotRewrite);
-        } else {
-            result = tr;
+        } else if (logic.isNot(symRef)) {
+            PTRef child = t[0];
+            PTRef newChild = cache[child];
+            result = child == newChild ? tr : logic.mkNot(newChild);
```

Comment:
I separated negation as the most common case (other than conjunction and disjunction) to try to avoid the temporary vector.